lib/rex/proto/kerberos/model/pre_auth_for_user.rb

```diff
@@ -0,0 +1,161 @@
+          end
+
+          # Encodes the user_realm attribute
+          #
+          # @return [OpenSSL::ASN1::GeneralString]
+          def encode_user_realm
+            OpenSSL::ASN1::GeneralString.new(user_realm)
+          end
+
+          # Encodes the cksum attribute
+          #
+          # @return [String]
+          def encode_cksum
+            cksum.encode
+          end
+
+          # Encodes the auth_package attribute
+          #
+          # @return [OpenSSL::ASN1::GeneralString]
+          def encode_auth_package
+            OpenSSL::ASN1::GeneralString.new(auth_package)
+          end
+
+          # Decodes a Rex::Proto::Kerberos::Model::PreAuthForUser from an String
+          #
+          # @param input [String] the input to decode from
+          # @raise [Rex::Proto::Kerberos::Model::Error::KerberosDecodingError] if decoding doesn't succeed
+          def decode_string(input)
+            asn1 = OpenSSL::ASN1.decode(input)
+
+            decode_asn1(asn1)
+          rescue OpenSSL::ASN1::ASN1Error
+            raise Rex::Proto::Kerberos::Model::Error::KerberosDecodingError
+          end
+
+          # Decodes a Rex::Proto::Kerberos::Model::PreAuthForUser from an
+          # OpenSSL::ASN1::Sequence
+          #
+          # @param input [OpenSSL::ASN1::Sequence] the input to decode from
+          # @raise [Rex::Proto::Kerberos::Model::Error::KerberosDecodingError] if decoding doesn't succeed
+          def decode_asn1(input)
+            seq_values = input.value
+
+            seq_values.each do |val|
+              case val.tag
+              when 0
+                self.user_name = decode_user_name(val)
+              when 1
+                self.user_realm = decode_user_realm(val)
+              #when 2
+              #  self.cksum = decode_cksum(val)
+              when 3
+                self.auth_package = decode_auth_package(val)
+              else
+                raise ::Rex::Proto::Kerberos::Model::Error::KerberosDecodingError, 'Failed to decode KdcRequestBody SEQUENCE'
+              end
+            end
+          end
+
+          # Decodes the user_name field
+          #
+          # @param input [OpenSSL::ASN1::ASN1Data] the input to decode from
+          # @return [Rex::Proto::Kerberos::Model::PrincipalName]
+          def decode_cname(input)
+            Rex::Proto::Kerberos::Model::PrincipalName.decode(input.value[0])
+          end
+
+          # Decodes the user_realm field
+          #
+          # @param input [OpenSSL::ASN1::ASN1Data] the input to decode from
+          # @return [String]
+          def decode_user_realm(input)
+            input.value[0].value
+          end
+
+          # Decodes the cksum field
+          #
+          # @param input [OpenSSL::ASN1::ASN1Data] the input to decode from
+          # @return [Rex::Proto::Kerberos::Model::PrincipalName]
+          #def decode_cksum(input)
+          #  Rex::Proto::Kerberos::Model::Checksum.decode(input.value[0])
+          #end
```

Comment:
Yes, you're right. I originally added the decoding logic of `PA-FOR-USER` to follow the way other similar structures are implemented. It is not used by the `get_ticket` module, which only needs to encode. I forgot to finish its implementation. I need to fix this since other modules might need to decode.

Reply:
Also, I'll add specs.